controllers/bucket_controller_test.go

```diff
@@ -17,59 +17,562 @@ limitations under the License.
 	"path/filepath"
+	"strings"
 	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/fluxcd/pkg/apis/meta"
+	"github.com/fluxcd/pkg/runtime/conditions"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
 )
 
-func TestBucketReconciler_checksum(t *testing.T) {
+func TestBucketReconciler_Reconcile(t *testing.T) {
+	g := NewWithT(t)
+
+	s3Server := newS3Server("test-bucket")
+	s3Server.Objects = []*s3MockObject{
+		{
+			Key:          "test.yaml",
+			Content:      []byte("test"),
+			ContentType:  "text/plain",
+			LastModified: time.Now(),
+		},
+	}
+	s3Server.Start()
+	defer s3Server.Stop()
+
+	g.Expect(s3Server.HTTPAddress()).ToNot(BeEmpty())
+	u, err := url.Parse(s3Server.HTTPAddress())
+	g.Expect(err).NotTo(HaveOccurred())
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "bucket-reconcile-",
+			Namespace:    "default",
+		},
+		Data: map[string][]byte{
+			"accesskey": []byte("key"),
+			"secretkey": []byte("secret"),
+		},
+	}
+	g.Expect(testEnv.Create(ctx, secret)).To(Succeed())
+	defer testEnv.Delete(ctx, secret)
+
+	obj := &sourcev1.Bucket{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "bucket-reconcile-",
+			Namespace:    "default",
+		},
+		Spec: sourcev1.BucketSpec{
+			Provider:   "generic",
+			BucketName: s3Server.BucketName,
+			Endpoint:   u.Host,
+			Insecure:   true,
+			Interval:   metav1.Duration{Duration: interval},
+			Timeout:    &metav1.Duration{Duration: timeout},
+			SecretRef: &meta.LocalObjectReference{
+				Name: secret.Name,
+			},
+		},
+	}
+	g.Expect(testEnv.Create(ctx, obj)).To(Succeed())
+
+	key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+
+	// Wait for finalizer to be set
+	g.Eventually(func() bool {
+		if err := testEnv.Get(ctx, key, obj); err != nil {
+			return false
+		}
+		return len(obj.Finalizers) > 0
+	}, timeout).Should(BeTrue())
+
+	// Wait for Bucket to be Ready
+	g.Eventually(func() bool {
+		if err := testEnv.Get(ctx, key, obj); err != nil {
+			return false
+		}
+		if !conditions.IsReady(obj) || obj.Status.Artifact == nil {
+			return false
+		}
+		readyCondition := conditions.Get(obj, meta.ReadyCondition)
+		return obj.Generation == readyCondition.ObservedGeneration &&
+			obj.Generation == obj.Status.ObservedGeneration
+	}, timeout).Should(BeTrue())
+
+	g.Expect(testEnv.Delete(ctx, obj)).To(Succeed())
+
+	// Wait for Bucket to be deleted
+	g.Eventually(func() bool {
+		if err := testEnv.Get(ctx, key, obj); err != nil {
+			return apierrors.IsNotFound(err)
+		}
+		return false
+	}, timeout).Should(BeTrue())
+}
+
+func TestBucketReconciler_reconcileStorage(t *testing.T) {
```

Comment:
`reconcileStorage()` implementations for bucket and gitrepository reconcilers seem to be exactly the same. Should we have a common function with this one test? Currently there's no test for `reconcileStorage` in gitrepository reconciler.

Reply:
I had noticed this too, but not taken the time yet to think about how I wanted to factor it out. The same goes for `garbageCollect` by the way.